### PR TITLE
[RFC] Make codec adapter a stand in layer

### PR DESCRIPTION
**Pull request by cujomalainey**
This PR is tackling the change of making codecs their own UUID rather than a sub id. Changes will come in gradually, but I wanted to get early feedback. Will add more stakeholders as larger changes come in.

### CODE REVIEW

File: src/include/sof/audio/codec_adapter/codec/generic.h [hunk 1]
```diff
@@ -25,6 +25,41 @@
 				(value)); \
 	} while (0)
 
+#define DECLARE_CODEC_ADAPTER(adapter, uuid, tr) \
+static struct comp_dev *adapter_shim_new(const struct comp_driver *drv, \
+					 struct sof_ipc_comp *comp)\
+{ \
+	return codec_adapter_new(drv, comp, adapter);\
+} \
+\
```
Comment (dbaluta) [inline src/include/sof/audio/codec_adapter/codec/generic.h:34]: Ok, I see where this is going. So, now we also need to modify codec_adapter_new to accommodate for the 3rd parameter. This is the way we tell the generic codec adapter which specific ops implementation to use.

Comment (cujomalainey) [reply src/include/sof/audio/codec_adapter/codec/generic.h:34]: yes, and we can add another hook into the register routine to register the codec to a codec list which the multiplexer can pull from

**Review by lgirdwood — APPROVED** [review]
LGTM

**Review by lgirdwood — APPROVED** [review]
@cujomalainey fwiw, I'm working in this area too and so will @ranj063 , lets have a sync up offline so we dont all do the same work.

**Comment by cujomalainey** [comment]
@lgirdwood agreed feel free to put something calendar and I will do my best to make room, mornings are usually the easiest for me though. I'm am stuck debugging this gem of an error though (im dumping pre processor output im so confused)

```
/home/sof/work/xtensa-cnl-elf/bin/xtensa-cnl-elf-gcc  -DRELATIVE_FILE=\"src/audio/codec_adapter/codec/passthrough.c\" -I/home/sof/work/sof.git/src/platform/intel/cavs/include -I/home/sof/work/sof.git/src/arch/xtensa/include -I/home/sof/work/sof.git/src/arch/xtensa/xtos -I/home/sof/work/sof.git/src/platform/tigerlake/include -I/home/sof/work/sof.git/src/platform/tigerlake/include/arch -I/home/sof/work/sof.git/../xtensa-root/xtensa-cnl-elf/include -I/home/sof/work/sof.git/src/include -I/home/sof/work/sof.git/rimage/src/include -I/home/sof/work/sof.git/build_tgl_gcc/generated/include  -nostdlib -fno-inline-functions -mlongcalls -O2 -g -Wall -Werror -Wl,-EL -Wmissing-prototypes -Wpointer-arith -mtext-section-literals -imacros/home/sof/work/sof.git/build_tgl_gcc/generated/include/autoconfig.h -o CMakeFiles/sof.dir/src/audio/codec_adapter/codec/passthrough.c.o   -c /home/sof/work/sof.git/src/audio/codec_adapter/codec/passthrough.c
/home/sof/work/xtensa-cnl-elf/bin/xtensa-cnl-elf-gcc  -DRELATIVE_FILE=\"src/lib/agent.c\" -I/home/sof/work/sof.git/src/platform/intel/cavs/include -I/home/sof/work/sof.git/src/arch/xtensa/include -I/home/sof/work/sof.git/src/arch/xtensa/xtos -I/home/sof/work/sof.git/src/platform/tigerlake/include -I/home/sof/work/sof.git/src/platform/tigerlake/include/arch -I/home/sof/work/sof.git/../xtensa-root/xtensa-cnl-elf/include -I/home/sof/work/sof.git/src/include -I/home/sof/work/sof.git/rimage/src/include -I/home/sof/work/sof.git/build_tgl_gcc/generated/include  -nostdlib -fno-inline-functions -mlongcalls -O2 -g -Wall -Werror -Wl,-EL -Wmissing-prototypes -Wpointer-arith -mtext-section-literals -imacros/home/sof/work/sof.git/build_tgl_gcc/generated/include/autoconfig.h -o CMakeFiles/sof.dir/src/lib/agent.c.o   -c /home/sof/work/sof.git/src/lib/agent.c
[ 77%] Building C object CMakeFiles/sof.dir/src/lib/lib.c.o
/home/sof/work/xtensa-cnl-elf/bin/xtensa-cnl-elf-gcc  -DRELATIVE_FILE=\"src/lib/lib.c\" -I/home/sof/work/sof.git/src/platform/intel/cavs/include -I/home/sof/work/sof.git/src/arch/xtensa/include -I/home/sof/work/sof.git/src/arch/xtensa/xtos -I/home/sof/work/sof.git/src/platform/tigerlake/include -I/home/sof/work/sof.git/src/platform/tigerlake/include/arch -I/home/sof/work/sof.git/../xtensa-root/xtensa-cnl-elf/include -I/home/sof/work/sof.git/src/include -I/home/sof/work/sof.git/rimage/src/include -I/home/sof/work/sof.git/build_tgl_gcc/generated/include  -nostdlib -fno-inline-functions -mlongcalls -O2 -g -Wall -Werror -Wl,-EL -Wmissing-prototypes -Wpointer-arith -mtext-section-literals -imacros/home/sof/work/sof.git/build_tgl_gcc/generated/include/autoconfig.h -o CMakeFiles/sof.dir/src/lib/lib.c.o   -c /home/sof/work/sof.git/src/lib/lib.c
In file included from /home/sof/work/sof.git/src/audio/codec_adapter/codec/passthrough.c:9:
/home/sof/work/sof.git/src/include/sof/audio/codec_adapter/codec/generic.h:53:27: error: expected ',' or ';' at end of input
   53 | static SHARED_DATA struct comp_driver_info comp_codec_adapter_info = { \
      |                           ^~~~~~~~~~~~~~~~
/home/sof/work/sof.git/src/audio/codec_adapter/codec/passthrough.c:99:1: note: in expansion of macro 'DECLARE_CODEC_ADAPTER'
   99 | DECLARE_CODEC_ADAPTER(passthrough_interface, passthrough_uuid, passthrough_tr)
      | ^~~~~~~~~~~~~~~~~~~~~
```

**Comment by lgirdwood** [comment]
@cujomalainey might be worth running the pre-processor only on that file and see the output if you think it's macro related.

**Comment by lgirdwood** [comment]
@cujomalainey thanks, yes, will schedule something next week with @ranj063 too.

**Comment by cujomalainey** [comment]
> @cujomalainey might be worth running the pre-processor only on that file and see the output if you think it's macro related.

Yep already doing that. Hopefully will have an answer soon.

**Comment by paulstelian97** [comment]
Please resubmit with "main" as PR base branch.

**Pull request closed without merging**